File: SAMER08D.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"
#define N 1007

int main(void) {
	ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
//    freopen("output.txt", "w", stdout);
#endif

    int arr[N][N], Count[N][N];
    for (int i = 0; i <= N; i++)
    	arr[i][0] = 0, arr[0][i] = 0;
    Count[0][0] = 0;
    int k;
    cin >> k;
    while (k) {
    	string s1, s2;
    	cin >> s1 >> s2;
    	int n = s1.size(), m = s2.size();
    	for (int i = 1; i <= n; i++) {
    		for (int j = 1; j <= m; j++) {
    			arr[i][j] = max(arr[i-1][j], arr[i][j-1]);
    			Count[i][j] = (s1[i-1] == s2[j-1]) ? Count[i-1][j-1] + 1 : 0;
    			if (Count[i][j] >= k)
    				for (int z = k; z <= Count[i][j]; z++)
    					arr[i][j] = max(arr[i][j], arr[i-z][j-z] + z);
    		}
    	}
    	cout << arr[n][m] << endl;
    	cin >> k;
    }
}

File: SAMER08D.py
N = 1007
arr = [[0] * N for i in range(N)]
Count = [[0] * N for i in range(N)]

while True:
	k = int(input())
	if k == 0:
		break
	s1 = input()
	s2 = input()
	for i in range(1, len(s1)+1):
		for j in range(1, len(s2)+1):
			arr[i][j] = max(arr[i-1][j], arr[i][j-1])
			if s1[i-1] == s2[j-1]:
				Count[i][j] = Count[i-1][j-1] + 1
			else:
				Count[i][j] = 0
			if Count[i][j] >= k:
				for z in range(k, Count[i][j]+1):
					arr[i][j] = max(arr[i][j], arr[i-z][j-z] + z)
	print(arr[len(s1)][len(s2)])
